### PR TITLE
Switch asymmetric encryption implementation to BoringSSL

### PR DESCRIPTION
**Pull request by wpaulino**
### What does this PR do?

We port over `crypto.publicEncrypt` and `crypto.privateDecrypt` as a pair as they are interdependent. All relevant test coverage from Node.js has been ported over, with some slight changes on some of the thrown error assertions.

I had the following benchmarks on an Apple M1 Pro:

This branch:

```
benchmark                              time (avg)             (min … max)       p75       p99      p999
------------------------------------------------------------------------- -----------------------------
RSA_PKCS1_OAEP_PADDING round-trip     829 µs/iter       (815 µs … 939 µs)    831 µs    883 µs    939 µs
```

bun/main:

```
benchmark                              time (avg)             (min … max)       p75       p99      p999
------------------------------------------------------------------------- -----------------------------
RSA_PKCS1_OAEP_PADDING round-trip  27'141 µs/iter (26'762 µs … 27'931 µs) 27'255 µs 27'931 µs 27'931 µs
```

Fixes https://github.com/oven-sh/bun/issues/4859.

### How did you verify your code works?

- [x] I included a test for the new code, or existing tests cover it
- [x] I ran my tests locally and they pass (`bun-debug test test-file-name.test`)

### CODE REVIEW

File: src/bun.js/bindings/KeyObject.cpp [hunk 1]
```diff
@@ -2943,6 +2947,150 @@ JSC_DEFINE_HOST_FUNCTION(KeyObject__SymmetricKeySize, (JSC::JSGlobalObject * glo
...
+    }
+
+    auto* jsKey = jsDynamicCast<JSObject*>(callFrame->uncheckedArgument(0));
+    if (!jsKey) {
+        return Bun::throwError(globalObject, scope, Bun::ErrorCode::ERR_INVALID_ARG_TYPE,
+            "expected object as first argument"_s);
+    }
+
+    auto jsCryptoKeyValue = jsKey->getIfPropertyExists(
+        globalObject, PropertyName(Identifier::fromString(vm, "key"_s)));
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/KeyObject.cpp:2968]: I think we can use `BuiltinNames` for this one:
```suggestion
        globalObject, builtinNames(vm).keyPublicName());
```

File: src/js/node/crypto.ts [hunk 2]
```diff
@@ -11610,16 +11612,10 @@ var require_privateDecrypt = __commonJS({
 var require_browser10 = __commonJS({
   "node_modules/public-encrypt/browser.js"(exports) {
     var publicEncrypt = require_publicEncrypt();
```
Comment (Jarred-Sumner) [inline src/js/node/crypto.ts:11614]: Can we delete the JS polyfill version or is there more to do before that?

File: src/bun.js/bindings/KeyObject.cpp [hunk 3]
```diff
@@ -2943,6 +2947,150 @@ JSC_DEFINE_HOST_FUNCTION(KeyObject__SymmetricKeySize, (JSC::JSGlobalObject * glo
     return JSC::JSValue::encode(JSC::jsUndefined());
 }
 
+static EncodedJSValue doAsymmetricCipher(JSGlobalObject* globalObject, CallFrame* callFrame, bool encrypt)
+{
+    auto count = callFrame->argumentCount();
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    if (count != 2) {
```
Comment (dylan-conway) [inline src/bun.js/bindings/KeyObject.cpp:2956]:
```suggestion
    if (count < 2) {
```
to be more javascripty

File: src/bun.js/bindings/KeyObject.cpp (hunk 1 above)
Comment (wpaulino) [reply src/bun.js/bindings/KeyObject.cpp:2968]:
```
error: no member named 'keyPublicName' in 'WebCore::BunBuiltinNames'; did you mean 'sepPublicName'?
 2967 |     auto jsCryptoKeyValue = jsKey->getIfPropertyExists(globalObject, builtinNames(vm).keyPublicName());
```

File: src/js/node/crypto.ts (hunk 2 above)
Comment (wpaulino) [reply src/js/node/crypto.ts:11614]: Once we port over `crypto.privateEncrypt/publicDecrypt` we can. I plan to tackle those next.

File: src/bun.js/bindings/KeyObject.cpp (hunk 3 above)
Comment (wpaulino) [reply src/bun.js/bindings/KeyObject.cpp:2956]: The crypto library is the one interfacing with this code though, so we can guarantee two arguments should always be passed in.

Comment (dylan-conway) [reply src/bun.js/bindings/KeyObject.cpp:2956]: ah, I see 👍

**Comment by Jarred-Sumner** [comment]
Thank you.

**Pull request merged**